add and export fun. optimize to use and overwrite defaults

samuelbelko opened the pull request:
When calling `optimize(f, inputdimension; optkwargs...) `, `optkwargs` overwrite and extend `optdefaults`, that overwrite and extend defaults in the constructor of `BOpt`. 
Maybe these defaults could be also placed somewhere else  in the future, in some kind of config file.

jbrea commented:
Great. Thanks.

I wonder, if it makes sense to give defaults for `lowerbounds` and `upperbounds` (and if defaults, than rather `-Inf` and `Inf`, although that would probably require some variable transformation internally). Maybe it would be better to just ask the user for `lowerbounds` and `upperbounds` or have only `f` as an argument and the user has to specify at least one of `lowerbounds`, `upperbounds` or `inputdimension`. What do you think?

jbrea commented:
Ah, and a second thought, for testing (and potentially for more clever default options in the future) it would probably be better to pull out the handling of options into a separate function `merge_with_defaults(optkwargs)` that returns `args` and `kwargs` and use this in `optimize`. (Btw. can one override `f` currently by giving `func` as an `optkwarg`; this shouldn't be possible). You could write a few simple tests to make sure `merge_with_defaults` behaves as it should.

samuelbelko commented:
Yes, I think you are right about giving defaults to `lowerbounds` and `upperbounds` is not optimal. These parameters are easy to understand by the user. They are very problem specific & heavily impact returned optimizers and so making them explicit is better. And plus we can eliminate `inputdimension` if the bounds are known. So I propose to make them not having any defaults.

Handling options in a seperate function is a very good idea. Also thanks for pointing out the bug with overwriting `f`, I will also add some tests 👍 .

DavidAfonsoValente commented:
You can consider adding a few more @assert statements to check the type and value of some arguments. For example, you can add assertions to check that inputdimension is an integer, lowerbounds and upperbounds are vectors, and the values of maxiterations, maxduration, etc., are non-negative.

DavidAfonsoValente commented:
Maybe instead of creating a tuple optdefaults for default arguments, you can consider creating a dictionary for easy extension and overwriting.

DavidAfonsoValente commented on src/BayesianOptimization.jl:
```diff
@@ -164,5 +225,84 @@ function _evaluate_function(o, x)
+        
+    inputdimension = length(lowerbounds)
+
+    all(lowerbounds .<= upperbounds) || 
```
Same here

samuelbelko commented:
> You can consider adding a few more @Assert statements to check the type and value of some arguments. For example, you can add assertions to check that inputdimension is an integer, lowerbounds and upperbounds are vectors, and the values of maxiterations, maxduration, etc., are non-negative.

Hi, thanks for your comments. My concern with `@assert` was that it might make sense to use tests for checking the configuration of default values, so that there is feedback after merging with master branch. (Do tests run automatically after merging?) I think that for issues caused by external data, exceptions should be thrown and asserts should be used for checking internal logic independent of external input.  I could add types to function signature to enforce `lowerbounds`, `upperbounds` to be vectors but I don't think it is very necessary. I added the check for `maxiterations >= 0` and `maxduration >=0` (using Exceptions) into BOpt constructor, also moved another one from `merge_with_defaults` there.

> Maybe instead of creating a tuple optdefaults for default arguments, you can consider creating a dictionary for easy extension and overwriting.

I think NamedTuples are just right. They are immutable, preserve order of keys and just model the behaviour of keyword arguments (I think that is why kwargs you get from a general function is a NamedTuple as well).

jbrea commented on src/BayesianOptimization.jl:
```diff
@@ -111,22 +158,28 @@ function show(io::IO, mime::MIME"text/plain", o::BOpt)
     xs = Vector{Float64}[]
     for x in o.initializer
-        for j in 1:o.repetitions
+        for j = 1:o.repetitions
```
This looks like the JuliaFormatters `DefaultStyle`. I'm not a big fan of the `for j = 1:n` style. Could we maybe use `SciMLStyle`?

samuelbelko replied:
Yes, of course. I will adjust it.

jbrea commented on src/BayesianOptimization.jl:
```diff
@@ -164,5 +232,80 @@ function _evaluate_function(o, x)
+"""
+function merge_with_defaults(f, lowerbounds, upperbounds, optkwargs)
+    # the same order of args as in BOpt constructor
+    args_keys = [:model, :acquisition, :modeloptimizer]
```
Use tuples `()` instead of vectors `[]` (they don't allocate at runtime).
Also for `kwargs_keys`.

samuelbelko replied:
oki

jbrea commented on src/BayesianOptimization.jl:
```diff
@@ -164,5 +225,84 @@ function _evaluate_function(o, x)
+    issubset(keys(optkwargs), union(args_keys, kwargs_keys)) ||
+        throw(ArgumentError("use of unsupported keyword arguments"))
+
+    length(lowerbounds) == length(upperbounds) ||
```
No. Throwing an `ArgumentError` is better.

jbrea replied:
No. Throwing and argument error is better.

jbrea commented on src/BayesianOptimization.jl:
```diff
@@ -164,5 +232,80 @@ function _evaluate_function(o, x)
+
+    # TODO: come up with reasonable optdefaults
+    # optdefaults can be extended to eventually overwrite default kwargs when passed to BOpt constructor
+    optdefaults = (
```
It could make sense to move the default options to a separate function `default_options(inputdimension)` and replace the `@assert`s with tests. But I don't have a strong opinion on this.

samuelbelko replied:
Okay cool, so I will keep it like it is for now.

jbrea commented on src/BayesianOptimization.jl:
```diff
@@ -164,5 +232,80 @@ function _evaluate_function(o, x)
+    # optkwargs overwrites and extends optdefaults
+    params = merge(optdefaults, optkwargs)
+    # split params into args and kwargs for BOpt constructor, args are in the right order for BOpt constructor
+    args = [f, [params[k] for k in args_keys]..., lowerbounds, upperbounds]
```
Use tuple instead of vector.

jbrea reviewed (commented):
Looks good overall, thanks!

jbrea commented:
Great. Thanks!
